File: Level 2/다리를 지나는 트럭/solution.js
function solution(bridge_length, weight, truck_weights) {
  let seconds = 0;
  let queue = [];

  while (queue.length > 0 || truck_weights.length > 0) {
    seconds++;
    queue = queue.map((truck) => ({
      ...truck,
      remainingSeconds: truck.remainingSeconds - 1,
    }));

    if (queue[0]?.remainingSeconds === 0) {
      queue.shift();
    }

    const isCanTruckOnBridge =
      queue.reduce((acc, { weight }) => acc + weight, 0) + truck_weights[0] <=
      weight;

    if (isCanTruckOnBridge) {
      const truckWeight = truck_weights.shift();
      queue.push({ weight: truckWeight, remainingSeconds: bridge_length });
      continue;
    }

    if (queue[0]?.remainingSeconds > 1) {
      const firstTruckRemainingSeconds = queue[0].remainingSeconds - 1;
      seconds += firstTruckRemainingSeconds;
      queue = queue.map((truck) => ({
        ...truck,
        remainingSeconds: truck.remainingSeconds - firstTruckRemainingSeconds,
      }));
    }
  }

  return seconds;
}

File: Level 2/다리를 지나는 트럭/solution.py
from collections import deque


def solution(bridge_length, weight, truck_weights):
    answer = 0

    bridge = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    total_weight = 0

    while truck_weights:
        total_weight -= bridge.popleft()

        if total_weight + truck_weights[0] <= weight:
            truck_weight = truck_weights.popleft()
            bridge.append(truck_weight)
            total_weight += truck_weight
        else:
            bridge.append(0)

        answer += 1

    answer += bridge_length

    return answer
